benchmarks/apps/sphng/sphng.py

```diff
@@ -0,0 +1,238 @@
+import os.path as path
+import reframe as rfm
+import reframe.core.runtime as rt
+import reframe.utility.sanity as sn
+from benchmarks.modules.utils import SpackTest
+
+
+def threads_per_part():
+    for p in rt.runtime().system.partitions:
+        nthr = 1
+        while nthr < p.processor.num_cores:
+            yield (p.fullname, nthr)
+            nthr <<= 1
+        yield (p.fullname, p.processor.num_cores)
+
+
+class SphngBase(SpackTest):
+
+    descr = 'Base class for Sphng'
+    valid_systems = ['*']
+    valid_prog_environs = ['default']
+    time_limit = '0d0h30m0s'
+    exclusive_access = True
+    spack_spec = 'sphng@v1.0.0%intel'
+    executable = 'sph_tree_rk_gradh'
+
+    num_tasks = required
+    num_tasks_per_node = required
+    num_cpus_per_task = required
+
+    def set_common_env_variables(self):
+        self.env_vars['OMP_PLACES'] = 'cores'
+
+        # pre-requisite IntelMPI's mpirun launcher so far...
+        if(self.current_partition.scheduler.registered_name == 'slurm'):
+            if(self.num_tasks_per_node <= self.current_partition.processor.num_cpus_per_socket):
+                self.env_vars['I_MPI_PIN_RESPECT_CPUSET'] = '0'
+        elif(self.current_partition.scheduler.registered_name == 'torque'):
+            self.env_vars['I_MPI_JOB_RESPECT_PROCESS_PLACEMENT'] = 'off'
+            self.executable = '-perhost $NUM_TASKS_PER_NODE ' + self.executable
+
+    @run_before('compile')
+    def setup_build_system(self):
+        self.build_system.specs = [self.spack_spec]
+
+
+class SphngBase_ifile(SphngBase):
+
+    prerun_cmds = ['rm -rf fort* TEST* notify ifile sphng_setup.o']
+    executable_opts = ['initial ifile < ./setup.txt &> sphng_setup.o']
+
+    @sanity_function
+    def validate_ifile_generation(self):
+        return sn.assert_true(path.exists('ifile'))
+
+
+class SphngBase_evolution(SphngBase):
+
+    executable_opts = ['evolution ifile']
+
+    reference = {
+        'dial:slurm-local': {
+            'Total elapsed time': (10, None, None, 'minutes'),
+            },
+        }
+
+    @run_before('sanity')
+    def run_complete_pattern(self):
+        self.pattern = r'ended on'
+        self.sanity_patterns = sn.assert_found(self.pattern, 'test01')
+
+    @performance_function('minutes')
+    def get_elapsed_time(self):
+        return sn.extractsingle(r'cpu time used for this run :\s+(\S+)\s', 'test01', 1, float)
+
+    @run_before('performance')
+    def runtime_extract_pattern(self):
+        self.perf_variables = {
+                'Total elapsed time':self.get_elapsed_time()
+                }
+
+
+class Sphng_Single_Node_ifile(SphngBase_ifile):
+
+    mpi_tasks = parameter(threads_per_part(), fmt=lambda x: x[1])
+
+    @run_after('init')
+    def setup_thread_config(self):
+        self.valid_systems = [self.mpi_tasks[0]]
```

Comment:
@tkoskela @giordano I have managed to use the feature they discussed to get mpi_tasks to be generated correctly for all partitions. To check I just made a dummy partition in the reframe_config with only 4 cpus and it generates the correct number of tests

Reply:
So I think we are finally ready to merge now :sweat_smile:

Reply:
This is probably a stupid question, but why does `self.mpi_tasks` go into `self.valid_systems`?

Reply:
It's not a stupid question :wink: 

I took this directly from [the example](https://github.com/reframe-hpc/reframe/blob/5e5e94c3655b58704f26725871ad42d4bd51d8e2/tutorials/cscs-webinar-2022/tests/stream9.py#L63-L75). 

`parameter(threads_per_part(), fmt=lambda x: x[1])` returns a tuple containing the system partition name and the number of mpi tasks. 

The issue as I understand is that the performance variable gets it's name from the parameter variable. So when test names and parameters are printed it will read as e.g., `mpi_tasks=4` due to the `fmt=lambda x: x[1]` part. I agree that the name is misleading but this is so that it fits nicely into the rest of reframe.